descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,308 @@
+                nrow = n//self.ncolumns
+                catalog_label = ' '.join([catalog_name, re.sub(self.band,filtername+' '+self.band,cutlabel)])
+                validation_label = ' '.join([self.validation_data.get('label', ''), cutlabel])
+                reskey = re.sub('\$','',cutlabel)
+                results[reskey] = self.make_subplot(meanz, z_this, n, ax_this, z0, z0err, catalog_color, validation_color, catalog_label, validation_label)
+                results[reskey]['total'] = total
+
+                #add curve for this catalog to summary plot
+                if self.first_pass:  #add validation data if evaluating first catalog
+                    summary = self.make_subplot(meanz, z_this, n, summary_ax_this, z0, z0err, catalog_color, validation_color, catalog_label, validation_label)
+                else:
+                    summary = self.make_subplot(meanz, z_this, n, summary_ax_this, 0., 0., catalog_color, validation_color, catalog_label, validation_label)
+            
+            else:
+                #make empty subplots invisible 
+                ax_this.set_visible(False)
+                summary_ax_this.set_visible(False)
+                
+        #save results for catalog and validation data in txt files
+        for filename, dtype, comment, info in zip_longest((filelabel, self.observation), ('y', 'fit'), (filtername,), ('total',)):
+            if len(filename)>0: 
+                with open(os.path.join(base_output_dir,'Nvsz_'+filename+'.txt'),'ab') as f_handle:     #open file in append mode
+                    #loop over magnitude cuts in results dict
+                    for key, value in results.items():
+                        self.save_quantities(dtype, value, f_handle, comment=' '.join(((comment or ''), key, value.get(info, ''))))
+
+        if self.first_pass: #turn off validation data plot in summary for remaining catalogs
+            self.first_pass = False
+
+        #save figure
+        plt.savefig(os.path.join(base_output_dir, 'Nvsz_'+filelabel+'.png'))
+        plt.close()
+        return TestResult(0, passed=True)
+
+    def make_subplot(self, meanz, catalog_data, nplot, f, z0, z0err, catalog_color, validation_color, catalog_label, validation_label):
+
+        results = {}
+        results['meanz'] = meanz
+        if nplot%self.ncolumns==0:  #1st column
+            f.set_ylabel('$'+self.yaxis+'$',size=fsize)
+
+        if nplot+1 <= self.nplots-self.ncolumns:  #x scales for last ncol plots only
+            #print "noticks",nplot
+            for axlabel in f.get_xticklabels():
+                axlabel.set_visible(False)
+                    
+                #prevent overlapping yaxis labels
+                f.yaxis.get_major_ticks()[0].label1.set_visible(False)
+        else:
+            f.set_xlabel('$z$',size=fsize)
+            for axlabel in f.get_xticklabels():
+                axlabel.set_visible(True)
+
+        #plot catalog data if available
+        if len(catalog_data)>0:
```

Comment:
`if catalog_data:`

Reply:
This gave me an error so I restored my original code. If we refactor make_subplot, we can deal with it then.